### PR TITLE
Circuit breaker pattern

### PR DESCRIPTION
**Pull request by geffzhang**
Refactor CircuitBreaker Pattern and Add Qos Config in ReRoute

### CODE REVIEW

**Comment by TomPallister** [comment]
@geffzhang code looks good!

I'm happy to merge this after I have finished the service discovery work.

**Comment by TomPallister** [comment]
Fixes #23 and is ok to merge

**Comment by geffzhang** [comment]
@TomPallister I merge the code ,but AcceptanceTests should_use_service_discovery_and_load_balance_request could not pass

**Comment by TomPallister** [comment]
@geffzhang I will pull the code and look at the tests.

**Comment by TomPallister** [comment]
@geffzhang the test is failing because CreateHttpMessageHandler() is broken when qos.ExceptionsAllowedBeforeBreaking is 0. I think Polly needs greater than 0.

Perhaps it would be better to add the QoS configuration to FileReRoute.cs if this is left null the user does not want to use QoS, else use the settings in the QoS configuration. So create a class like below.

public class FileQoS
{
       public int ExceptionsAllowedBeforeBreaking { get; set; }

        public TimeSpan DurationOfBreak { get;set; }

        public TimeSpan TimeoutValue { get; set; }

        public TimeoutStrategy TimeoutStrategy { get; set; }
}

Then add Qos property to ReRoute, change constructor to take QoS and expose public Getter.

This is then used in FileOcelotConfigurationCreator.SetUpReRoute(FileReRoute fileReRoute, FileGlobalConfiguration globalConfiguration)

We map FileQos to Qos and populate the ReRoute QoS parameter.

Then when we build the Request in HttpRequestCreator we check to see if we should use QoS. If true call WithQos else don't.

Then in HttpClientBuilder we check if we should use QosS before adding a handler.

QoS should be turned off by default and the user can have the option to turn it on?

**Comment by geffzhang** [comment]
@TomPallister I Refactor qos as options，But  another test should_use_global_request_id_and_forward not pass

**Comment by TomPallister** [comment]
@geffzhang I will check again when I get home from work :)

File: test/Ocelot.AcceptanceTests/RequestIdTests.cs [hunk 1]
```diff
@@ -95,8 +95,7 @@ public void should_use_global_request_id_and_forward()
                             DownstreamPort = 51879,
                             DownstreamScheme = "http",
                             DownstreamHost = "localhost",
-                            UpstreamTemplate = "/",
-                            UpstreamHttpMethod = "Get",
+                            UpstreamTemplate = "/",
```
Comment (TomPallister) [inline test/Ocelot.AcceptanceTests/RequestIdTests.cs]: @geffzhang This test needs UpstreamHttpMethod = "Get",

**Comment by TomPallister** [comment]
@geffzhang RequestIdTests.should_use_global_request_id_and_forward needs UpstreamHttpMethod = "Get" adding to FileReRoute

**Comment by TomPallister** [comment]
@geffzhang thank you for helping with Ocelot.

**Comment by rahulbhat13** [comment]
@TomPallister @geffzhang I am new to ocelot , i want to use the x-forwaded filters ,qos and circuit breaker to pivotal cloud foundry so can you please suggest about this .
Qos is just a Json config i believe , x-forwaded i have no idea as of now and how polly will help in pcf . Please suggest

**Pull request merged**